### PR TITLE
Upgrading versions of used frameworks

### PR DESCRIPTION
**Pull request by jchlebik**
Upgrading versions of used frameworks:
  - Python 3.9.6.
  - CUDA 11.1
  - PyTorch 1.9.0
  - PyTorch Lightning 1.4.0

Provided a new environment file

Updating spectral module as the original fft calls were removed around 1.6 in favor of fft modules with a different API

Reimagining source as a module instead of creating new instances. This source module is using torch instead of numpy for the math calls

Check for capacity in replay buffer

train.py script edited to comply with newer checkpointing callbacks

cut out the genetic search part, let me know if people are interested in adding it

BUG/ISSUE: random cuda memory corruption exception after the lightning threads get killed

### CODE REVIEW

**Comment by astanziola** [comment]
Thanks @jchlebik ! The PR looks excellent,  I'll give it a look and let you know if any comments.
I used to have the same memory corruption problem with the old code, when stopping the training halfway.

About the genetic search part, I think it is pretty much up to you whether to publish it or not. But surely I would find it interesting :smile:

File: helmnet/hybridnet.py [hunk 1]
```diff
@@ -73,21 +82,38 @@ def init_f(self):
...
-                state_depth=self.hparams.state_depth,
+                activation_function = self.hparams.activation_function,
+                depth               = self.hparams.depth,
+                domain_size         = self.hparams.domain_size,
+                features            = self.hparams.features,
+                inchannels          = 6,
+                state_channels      = self.hparams.state_channels,
+                state_depth         = self.hparams.state_depth,
             )
+        # elif nn_name == "genetic_unet":
```
Comment (astanziola) [inline helmnet/hybridnet.py:93]: shall we just take those various pieces of commented code out, until we add the code for the genetic algorithms?

File: helmnet/hybridnet.py [hunk 2]
```diff
@@ -220,6 +251,17 @@ def val_dataloader(self):
             num_workers=min([self.hparams.batch_size, 32]),
         )
 
+    def test_dataloader(self):
+        self.reset_source()
+        testset = get_dataset("datasets/splitted_96/testset.ph")
```
Comment (astanziola) [inline helmnet/hybridnet.py:256]: Should we pass this path to the `IterativeSolver` from the `settings`?  I think that different people may use different datasets, possibly in the same project

File: helmnet/hybridnet.py [hunk 3]
```diff
@@ -276,13 +320,31 @@ def test_step(self, batch, batch_idx):
...
                 return_states=False,
             )
-            # Get loss
-            losses = [self.test_loss_function(x) for x in output["residuals"]]
-            losses = torch.stack(losses, 1)
-            return {
-                "losses": losses,
-                "wavefields": [x.cpu() for x in output["wavefields"]],
-            }
+        # Get loss
```
Comment (astanziola) [inline helmnet/hybridnet.py:323]: Should this be intented, to stay in the scope of `torch.no_grad()`, or perhaps lightning already takes care of this?

File: helmnet/spectral.py [hunk 4]
```diff
@@ -29,26 +29,45 @@ def conj(x):
...
+            dim = (-2, -1), 
+            norm = "backward"
+        )
+    )
 
     # get derivatives
-    dx = complex_mul(u_fft, kx)
-    dy = complex_mul(u_fft, ky)
+    dx  = complex_mul(u_fft, kx)
+    dy  = complex_mul(u_fft, ky)
```
Comment (astanziola) [inline helmnet/spectral.py:52]: Those `complex_mul` and `conj` functions were used because pytorch didn't support backpropagation trough operations with complex numbers.
Do we still need them, or can pytorch now work with complex values and we can replace them with built-in functions?

File: helmnet/spectral.py [hunk 5]
```diff
@@ -129,11 +148,17 @@ def __init__(self, size: int, direction="x"):
 
     def forward(self, x):
         """x must be [batch, x, y, real/imag]"""
-        return torch.ifft(
-            torch.fft(x, signal_ndim=2, normalized=False).flip(dims=[3]).mul(self.k),
-            signal_ndim=2,
-            normalized=False,
-        )
+        return torch.view_as_real(torch.fft.ifftn(
```
Comment (astanziola) [inline helmnet/spectral.py:151]: Same here. If it is easier to work with complex numbers we could switch to them, and we convert to/from real tensors with two channels only when interfacing with the neural network

**Review by astanziola — COMMENTED** [review]
Thanks again for the PR, that's an awesome amount of work! 
I gave it a try, and there are a few deprecation warnings that I'm personally not worried about, however I do get this when the training starts:

```
Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any
unused parameters in the forward pass. This flag results in an extra traversal of the autograd
graph every iteration,  which can adversely affect performance. If your model indeed never has
any unused parameters in the forward pass, consider turning this flag off. Note that this
warning may be a false positive if your model has flow control causing later iterations to have
unused parameters. (function operator())
```
not sure how much performances are really affected. Any ideas?

File: helmnet/hybridnet.py [hunk 6]
```diff
@@ -449,13 +505,21 @@ def training_step(self, sos_batch, batch_idx):
                     )
                     terminal_logged = True
 
+        self.logger.experiment.add_scalar(
+            "val_loss",
+            loss,
+            self.trainer.global_step
+        )
+
+        self.log('val_loss', loss) # force the variable to be saved in state dict of module. tensorboard logger does not do this for some reason
```
[the remark this answers is not in the document]
Comment (jchlebik) [reply helmnet/hybridnet.py:514]: It probably should. Checkpointer uses this metric as a monitor, is that also wrong or no?

File: helmnet/spectral.py (hunk 4 above)
Comment (jchlebik) [reply helmnet/spectral.py:52]: According to docs, pytorch autograd now works with complex operations, however JIT subsystem is still not supported. https://pytorch.org/docs/stable/complex_numbers.html

File: helmnet/hybridnet.py (hunk 3 above)
Comment (jchlebik) [reply helmnet/hybridnet.py:323]: Seems so, will remove in next commit. https://github.com/PyTorchLightning/pytorch-lightning/issues/2171

File: helmnet/hybridnet.py (hunk 2 above)
Comment (jchlebik) [reply helmnet/hybridnet.py:256]: Sounds great, will include in next commit.

File: helmnet/hybridnet.py (hunk 1 above)
Comment (jchlebik) [reply helmnet/hybridnet.py:93]: Yes, will remove it in next commit. On a side note, im going to create another branch with the full genetic search part included. That part will need more documentation before merging into master.

Comment (astanziola) [reply helmnet/hybridnet.py:93]: Sounds great! To be honest, the whole code needs more documentation, I sit down one day and do so!

File: helmnet/hybridnet.py (hunk 3 above)
Comment (astanziola) [reply helmnet/hybridnet.py:323]: Oh ok, if it does then is probably right as it is!

File: helmnet/hybridnet.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (astanziola) [reply helmnet/hybridnet.py:514]: I think that the checkpointer should save the model with the best validation loss, so shouldn't it be checked in the validation loop, rather than the training loop?

**Comment by astanziola** [comment]
Hi @jchlebik 
Sorry if it took a while, but I finally managed to review the PR. 
That's a huge amount of work, thank you very much!
